### PR TITLE
Plutopulp/default and public environments

### PR DESCRIPTION
**Pull request by plutopulp**
# Pull request outline

In this PR we we add to the environments CLI, where we:
- Refactor existing logic for CRUD by_name on environments by resolving/fetching the environment right off the bat, by `name` or `id` and then use the ID of the resulting object to make further actions on that environment.
- Add logic for listing publicly available environments, e.g. `pipeline environments list --public`
- Add logic for retrieving the default environment, e.g. `pipeline environments get --default`

Note that the default environment configuration `(id, name)` is always expected to be defined in this client, i.e. `pipeline-ai` and not in the `API` itself, even if the environment object associated with this configuration, is defined in the API. This lets the client choose what the default environment should be. Just something to note.
Further, the way we have implemented getting the default environment isn't the prettiest, where we conditionally add the required parser argument `name_or_id` depending on whether the `--default` flag has been set. This makes the UX of the CLI potentially nicer, but the following for instance, would keep things less convoluted: `pipeline environments get-default`. 

## Checklist:
- [ ] Docs updated
- [x] Tests written
- [x] Check for breaking changes in Resource
- [x] Version bumped

___

- _You can add a reference to an issue using hash followed by the issue number_
- _If a checklist item is to be ignored it must be struck through_

### CODE REVIEW

File: pipeline/console/__init__.py [hunk 1]
```diff
@@ -282,9 +282,16 @@ def main(args: Optional[List[str]] = None) -> int:
...
-
     environments_get_parser.add_argument(
-        "name_or_id", help="The environment name or id to get"
+        "--default",
+        help="Get the default environment",
+        action="store_true",
+    )
+    # Positional arguments are required. We don't want to have to include name_or_id
+    # when the default flag is set
+    "--default" not in sys.argv and environments_get_parser.add_argument(
```
Comment (alexpearce) [inline pipeline/console/__init__.py:292]: I'd prefer `if "--default" not in sys.argv:` but won't insist.

File: pipeline/console/environments.py [hunk 2]
```diff
@@ -18,37 +19,46 @@
...
     remote_service = PipelineCloud(verbose=False)
     remote_service.authenticate()
 
-    url = (
-        f"/v2/environments/by-name/{name_or_id}"
-        if by_name
-        else f"/v2/environments/{name_or_id}"
-    )
-
+    url = get_url(name_or_id, by_name, default)
```
Comment (alexpearce) [inline pipeline/console/environments.py:41]: Ahh, if only Python `if`s were expressions…

File: pipeline/console/environments.py [hunk 3]
```diff
@@ -84,87 +94,75 @@ def _create_environment(name: str, packages: List[str]) -> EnvironmentGet:
...
-        f"/v2/environments/{environment_information.id}",
+        f"/v2/environments/{id}",
         update_schema.dict(),
     )
     return EnvironmentGet.parse_obj(response)
 
 
 def _add_packages_to_environment(
-    name_or_id: str,
+    environment: EnvironmentGet,
```
Comment (alexpearce) [inline pipeline/console/environments.py:132]: I like the way you pass in this environment object!

**Review by alexpearce — DISMISSED** [review]
Nice refactor and nice tests to go with it. Thank you!

File: pipeline/console/__init__.py (hunk 1 above)
Comment (plutopulp) [reply pipeline/console/__init__.py]: Yeah I hear you ;) Updated. Thx

File: pipeline/console/environments.py (hunk 3 above)
Comment (plutopulp) [reply pipeline/console/environments.py:132]: Yeah, made sense since we're fetching it off the bat and needed other non-id properties in the method.

**Review by alexpearce — APPROVED** [review]
🚢

**Pull request merged**